cpp/src/arrow/flight/internal.cc

```diff
@@ -49,27 +50,117 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
     return Status::OK();
   }
 
-  if (grpc_status.error_code() == grpc::StatusCode::UNIMPLEMENTED) {
-    return Status::NotImplemented("gRPC returned unimplemented error, with message: ",
+  switch (grpc_status.error_code()) {
+    case grpc::StatusCode::OK:
+      return Status::OK();
+    case grpc::StatusCode::CANCELLED:
+      return Status::IOError("gRPC cancelled call, with message: ",
+                             grpc_status.error_message())
+          .WithDetail(std::make_shared<FlightStatusDetail>(FlightStatusCode::Cancelled));
+    case grpc::StatusCode::UNKNOWN:
+      return Status::UnknownError("gRPC returned unknown error, with message: ",
                                   grpc_status.error_message());
-  } else {
-    return Status::IOError("gRPC failed with error code ", grpc_status.error_code(),
-                           " and message: ", grpc_status.error_message());
+    case grpc::StatusCode::INVALID_ARGUMENT:
+      return Status::Invalid("gRPC returned invalid argument error, with message: ",
+                             grpc_status.error_message());
+    case grpc::StatusCode::DEADLINE_EXCEEDED:
+      return Status::IOError("gRPC returned deadline exceeded error, with message: ",
+                             grpc_status.error_message())
+          .WithDetail(std::make_shared<FlightStatusDetail>(FlightStatusCode::TimedOut));
+    case grpc::StatusCode::NOT_FOUND:
+      return Status::KeyError("gRPC returned not found error, with message: ",
+                              grpc_status.error_message());
+    case grpc::StatusCode::ALREADY_EXISTS:
+      return Status::AlreadyExists("gRPC returned already exists error, with message: ",
+                                   grpc_status.error_message());
+    case grpc::StatusCode::PERMISSION_DENIED:
+      return Status::IOError("gRPC returned permission denied error, with message: ",
+                             grpc_status.error_message())
+          .WithDetail(
+              std::make_shared<FlightStatusDetail>(FlightStatusCode::Unauthorized));
+    case grpc::StatusCode::RESOURCE_EXHAUSTED:
+      return Status::Invalid("gRPC returned resource exhausted error, with message: ",
+                             grpc_status.error_message());
+    case grpc::StatusCode::FAILED_PRECONDITION:
```

Comment:
FAILED_PRECONDITION, ABORTED, and RESOURCE_EXHAUSTED  errors might be good candidates to make top-level.  We should probably discuss on the mailing list.  But they have slightly different semantics and might be useful to build more general purpose machinery around handling them